[webkitscmpy] Generate commit message template

JonWBedard opened the pull request:
#### 7dfc51593d68bc49bec3d73231f0171c3abf8139
<pre>
[webkitscmpy] Generate commit message template
<a href="https://bugs.webkit.org/show_bug.cgi?id=231023">https://bugs.webkit.org/show_bug.cgi?id=231023</a>
&lt;rdar://problem/83722871 &gt;

Reviewed by NOBODY (OOPS!).

* Tools/Scripts/git-webkit: Use prepare-ChangeLog to generate commit messages.
* Tools/Scripts/hooks/prepare-commit-msg: Added.
* Tools/Scripts/libraries/webkitscmpy/setup.py: Bump version, add jinja.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/__init__.py: Ditto.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/mocks/local/git.py:
(Git.__init__): Add unbound status mock.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/__init__.py:
(main): Pass hooks to sub-programs.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/setup.py:
(Setup.git): iterate through the directory of provided hooks and render with jinja2.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/hooks/prepare-commit-msg: Added.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/setup_unittest.py:
(TestSetup.test_commit_message):
</pre>

huskyroll reviewed (changes requested):
Make sure we align our internal markings for the review needing changes with GitHubs

kocsenc commented on Tools/Scripts/git-webkit:
```diff
@@ -45,6 +46,43 @@ def is_webkit_filter(to_return):
+    import subprocess
+
+    dirname = None
+    content = []
```
Could probably be more descriptive here. `commit_message` maybe?

kocsenc commented on Tools/Scripts/git-webkit:
```diff
@@ -45,6 +46,43 @@ def is_webkit_filter(to_return):
+            if line == '~':
+                dirname = None
+                continue
+            if line.startswith(' ' * 8):
```
Not sure if worth breaking out that 8 into it's variable since it's used a couple times here.

kocsenc commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/setup.py:
```diff
@@ -29,6 +32,41 @@
+def prepare_commit_message_main(git, generator, file_name=None, source=None, sha=None):
+    if source and source != 'commit':
+        return 0
+    if not file_name:
```
nit: add whitespace between the `if` statements because there is a `return` above. This is 100% style/preference but I think it helps isolate the code as a return conditional.

kocsenc commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/setup.py:
```diff
@@ -29,6 +32,41 @@
+    if source and source != 'commit':
+        return 0
+    if not file_name:
+        sys.stderr.write('No commit message file provided to prepare-commit-msg\n')
```
I doubt this would ever happen right? git will _always_ guarantee a file to write on, even if it's empty.

kocsenc commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/setup.py:
```diff
@@ -29,6 +32,41 @@
+        return 0
+    if not file_name:
+        sys.stderr.write('No commit message file provided to prepare-commit-msg\n')
+        return 1
```
Ditto whitespace style nit.

kocsenc commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/setup.py:
```diff
@@ -29,6 +32,41 @@
+    if not file_name:
+        sys.stderr.write('No commit message file provided to prepare-commit-msg\n')
+        return 1
+    with open(file_name, 'w') as file:
```
code smell to use keyword `file` as a variable name. Maybe just `f`? or `commit_message_file` if we wanna go more verbose.

kocsenc commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/setup.py:
```diff
@@ -141,6 +179,41 @@ def git(cls, args, repository, **kwargs):
+        function_name = 'prepare_commit_message'
+        if prepare_commit_message and inspect.isfunction(prepare_commit_message) and prepare_commit_message.__name__:
+            function_name = prepare_commit_message.__name__
+            prepare_commit_message = inspect.getsource(prepare_commit_message)
```
So, although clever, what benefits are we getting from having the `prepare-commit-msg` script a function as opposed to a standalone file?

JonWBedard replied:
Most of the time projects are defining `prepare-commit-msg`, they're only doing so for one repository. This means that you can take advantage of the fact that you know how your repository is setup up and that `.git/hooks` is in a predictable location relative to your scripts. The problem we have in `git-webkit` is that we want it to work regardless of the repository we're targeting, which means that we know where `webkitscmpy` is when we run `git-webkit setup`, but _not_ when we actually run `prepare-commit-msg`. If we generate `prepare-commit-msg` like this, it lets us embed the known location of things in `prepare-commit-msg` based on the content of `git-webkit setup`. At the moment, we're only leveraging this in `Tools/Scripts/git-webkit`, but I suspect we will leverage it here in the near future when (hopefully) `webkitscmpy` owns commit message generation directly instead of relying on a perl script.

JonWBedard replied:
Worth calling out the other two approaches here, though: jinja templates or laying down a config file in `hooks` with information linking back to the location `git-webkit setup` is called from

kocsenc commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/setup.py:
```diff
@@ -141,6 +179,41 @@ def git(cls, args, repository, **kwargs):
+
+            # This is a bit hacky, but it allows the definition of prepare_commit_message to be owned by the caller
+            pcm_file = os.path.join(repository.root_path, '.git', 'hooks', 'prepare-commit-msg')
+            if not os.path.exists(os.path.dirname(pcm_file)):
```
If the `.git/hooks` path doesn't exist it means something else is wrong - we may be hiding other issues here but probably OK to leave.

kocsenc commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/setup.py:
```diff
@@ -141,6 +179,41 @@ def git(cls, args, repository, **kwargs):
+            pcm_file = os.path.join(repository.root_path, '.git', 'hooks', 'prepare-commit-msg')
+            if not os.path.exists(os.path.dirname(pcm_file)):
+                os.makedirs(os.path.dirname(pcm_file))
+            with open(pcm_file, 'w') as file:
```
Ditto re: `file` -> `f`

kocsenc reviewed (approved):
This looks clean. It's easier to follow than the previous version with not many drawbacks. r=me

huskyroll reviewed (approved):
Fix the conflicts and we can land it!

JonWBedard committed:
[webkitscmpy] Generate commit message template
https://bugs.webkit.org/show_bug.cgi?id=231023
<rdar://problem/83722871>

Reviewed by NOBODY (OOPS!).

* Tools/Scripts/git-webkit: Use prepare-ChangeLog to generate commit messages.
* Tools/Scripts/hooks/prepare-commit-msg: Added.
* Tools/Scripts/libraries/webkitscmpy/setup.py: Bump version, add jinja.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/__init__.py: Ditto.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/mocks/local/git.py:
(Git.__init__): Add unbound status mock.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/__init__.py:
(main): Pass hooks to sub-programs.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/setup.py:
(Setup.git): iterate through the directory of provided hooks and render with jinja2.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/hooks/prepare-commit-msg: Added.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/setup_unittest.py:
(TestSetup.test_commit_message):

dewei-zhu reviewed (approved):
r=me
